### ISSUE TITLE
Albanodecontroller purges before safety checking

### ISSUE THREAD

**Issue by JeffreyDevloo** [openvstorage/framework#1018]
## Problem description

When running some benchmark tests I encountered the following error:

```
 u'zNu0OORkDWBopYIBLsGOLgMO19lTtHPk', 'expected_safety': {u'good': 0, u'critical': 1, u'lost': 0}}
2016-10-11 11:23:28 83900 +0200 - cmp04 - 29569/140511254132544 - lib/albanode - 8375 - DEBUG - Removing ASD zNu0OORkDWBopYIBLsGOLgMO19lTtHPk at node 172.19.5.30
2016-10-11 11:23:29 13500 +0200 - cmp04 - 29569/140511254132544 - lib/albanode - 8376 - DEBUG - Purging ASD zNu0OORkDWBopYIBLsGOLgMO19lTtHPk on backend 5cee3299-1cb1-4e22-9f4a-dd148d637715
2016-10-11 11:23:29 18900 +0200 - cmp04 - 29569/140511254132544 - lib/albanode - 8377 - DEBUG - Removing ASD zNu0OORkDWBopYIBLsGOLgMO19lTtHPk from disk ata-SAMSUNG_MZ7LM480HCHP-00003_S1YJNX0H801485
2016-10-11 11:23:29 26200 +0200 - cmp04 - 29569/140511254132544 - lib/scheduled tasks - 8378 - ERROR - Ensure single CHAINED mode - ID 1476177808_w488W2jVBl - Task albanode.remove_asd with params {'node_guid': 'e589df0b-751e-41c7-a624-4178ffd06db8', 'asd_id': u'zNu0OORkDWBopYIBLsGOLgMO19lTtHPk', 'expected_safety': {u'good': 0, u'critical': 1, u'lost': 0}} failed
RuntimeError: Cannot remove ASD zNu0OORkDWBopYIBLsGOLgMO19lTtHPk as the current safety is not as expected ({'good': 0, 'critical': 0, 'lost': 1} vs {u'good': 0, u'critical': 1, u'lost': 0})
2016-10-11 11:23:29 28100 +0200 - cmp04 - 16755/140511254132544 - celery/celery.worker.job - 8364 - ERROR - Task albanode.reset_asd[5d3fc9b0-7119-4172-ab37-92bc393861cd] raised unexpected: RuntimeError("Cannot remove ASD zNu0OORkDWBopYIBLsGOLgMO19lTtHPk as the current safety is not as expected ({'good': 0, 'critical': 0, 'lost': 1} vs {u'good': 0, u'critical': 1, u'lost': 0})",)
RuntimeError: Cannot remove ASD zNu0OORkDWBopYIBLsGOLgMO19lTtHPk as the current safety is not as expected ({'good': 0, 'critical': 0, 'lost': 1} vs {u'good': 0, u'critical': 1, u'lost': 0})
```

This caused my asd to be unavailable and leaving me with a broken disk that could no longer be re-initialized.

The logs tell us that the purge (removal of the alba entry) is happening before the check whether the disk may be purged has been concluded. This means that my data would be discarded whether or not the disk is safe to delete (via web-api, GUI will prompt for extra confirmation)
## Possible root of the problem

The flow in the code is wrong
## Possible solution

Albanodecontroller: line 212-145 should be swapped with 217-230
## Temporary solution

None
### Setup

Hyperconverged setup
- Two nodes with each six disks for the back-end
### Package information
- openvstorage                          2.7.4-rev.4110.61da5ac-1                                 amd64        openvStorage
- openvstorage-backend                  1.7.4-rev.738.c050b22-1                                  amd64        openvStorage Backend plugin
- openvstorage-backend-core             1.7.4-rev.738.c050b22-1                                  amd64        openvStorage Backend plugin core
- openvstorage-backend-webapps          1.7.4-rev.738.c050b22-1                                  amd64        openvStorage Backend plugin Web Applications
- openvstorage-cinder-plugin            1.2.2-rev.38.dcc3b76-1                                   amd64        OpenvStorage Cinder plugin for OpenStack
- openvstorage-core                     2.7.4-rev.4110.61da5ac-1                                 amd64        openvStorage core
- openvstorage-hc                       1.7.4-rev.738.c050b22-1                                  amd64        openvStorage Backend plugin HyperConverged
- openvstorage-sdm                      1.6.4-rev.380.cfbfc0a-1                                  amd64        Open vStorage Backend ASD Manager
- openvstorage-webapps                  2.7.4-rev.4110.61da5ac-1                                 amd64        openvStorage Web Applications

**Comment by khenderick** [comment]
First, check for safety, then purge the OSD.

**Comment by khenderick** [comment]
Fixed by openvstorage/framework-alba-plugin#238, packaged in openvstorage-backend-1.7.4-rev.748.59bc937

**Comment by JeffreyDevloo** [comment]
### Steps
- Updated the environment where the issue occurred
- Ran the same benchmark suite
- Parsed log for safety errors

### Output

Found the following entry:

```
2016-10-11 21:31:37 47900 +0200 - cmp01 - 11671/140064603072320 - lib/albanode - 2559 - DEBUG - Removing ASD 8wHNffXFOWEHHtwOL6VpJA7udcZiEu3A at node 172.19.5.30
2016-10-11 21:31:38 35400 +0200 - cmp01 - 11671/140064603072320 - lib/scheduled tasks - 2560 - ERROR - Ensure single CHAINED mode - ID 1476214297_pR1LBDBO6D - Task albanode.remove_asd with params {'node_guid': '7a200087-26d7-4938-8
9a3-c18ead436870', 'asd_id': u'8wHNffXFOWEHHtwOL6VpJA7udcZiEu3A', 'expected_safety': {u'good': 0, u'critical': 1, u'lost': 0}} failed
Traceback (most recent call last):
  File "/opt/OpenvStorage/ovs/lib/helpers/decorators.py", line 305, in new_function
    output = function(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/albanodecontroller.py", line 220, in remove_asd
    raise RuntimeError('Cannot remove ASD {0} as the current safety is not as expected ({1} vs {2})'.format(asd_id, final_safety, expected_safety))
RuntimeError: Cannot remove ASD 8wHNffXFOWEHHtwOL6VpJA7udcZiEu3A as the current safety is not as expected ({'good': 0, 'critical': 0, 'lost': 1} vs {u'good': 0, u'critical': 1, u'lost': 0})
2016-10-11 21:31:38 35600 +0200 - cmp01 - 11671/140064603072320 - lib/scheduled tasks - 2561 - INFO - Ensure single CHAINED mode - ID 1476214297_pR1LBDBO6D - Amount of jobs pending for key ovs_ensure_single_albanode.remove_asd: 0
2016-10-11 21:31:38 36400 +0200 - cmp01 - 15218/140064603072320 - celery/celery.worker.job - 2547 - ERROR - Task albanode.reset_asd[2e26d9d4-0511-4df8-a888-e46bc8d8ba6a] raised unexpected: RuntimeError("Cannot remove ASD 8wHNffXFOW
EHHtwOL6VpJA7udcZiEu3A as the current safety is not as expected ({'good': 0, 'critical': 0, 'lost': 1} vs {u'good': 0, u'critical': 1, u'lost': 0})",)
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/albanodecontroller.py", line 264, in reset_asd
    disk_id = AlbaNodeController.remove_asd(node_guid, asd_id, expected_safety)
  File "/usr/local/lib/python2.7/dist-packages/celery/local.py", line 188, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 439, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/helpers/decorators.py", line 305, in new_function
    output = function(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/albanodecontroller.py", line 220, in remove_asd
    raise RuntimeError('Cannot remove ASD {0} as the current safety is not as expected ({1} vs {2})'.format(asd_id, final_safety, expected_safety))

```

There is no log indicating a purge like in the issue report.

### Test result

Test passed.
